ecs-cli/modules/clients/aws/route53/route53_client.go

```diff
@@ -0,0 +1,35 @@
+// Copyright 2015-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+// Package route53 contains functions for working with the route53 APIs
+// that back ECS Service Discovery
+package route53
+
+import (
+	"github.com/aws/amazon-ecs-cli/ecs-cli/modules/clients"
+	"github.com/aws/amazon-ecs-cli/ecs-cli/modules/config"
+	"github.com/aws/aws-sdk-go/service/route53"
+)
+
+// Private Route53 Client that can be mocked in unit tests
+// The SDK's route53 client implements this interface
+type route53Client interface {
+	GetHostedZone(input *route53.GetHostedZoneInput) (*route53.GetHostedZoneOutput, error)
+}
+
+// mockable factory function to create clients
+var newRoute53Client = func(config *config.CommandConfig) route53Client {
```

Comment:
[multiple occurrences] I'd very much highly encourage you to not follow the pattern of overriding global variables in tests. Speaking from ECS agent experience, global variables, especially those that get overridden in tests lead to all kinds of issues, in terms of complexity of code (as is apparent by the number of conversations on this PR 😃), maintainability and predictability. They also make tests stateful where, if one test doesn't properly clean up after itself, the next test fails.

Reply:
Alright, its pretty clear to me that no one likes the way I have used first class functions in this code; I'll change it to remove this function and the other one. I still standby the design idea I explained in [this comment](https://github.com/aws/amazon-ecs-cli/pull/565/#discussion_r208128647), but luckily I've come up with a way to keep that and also get rid of this function. 

However, I think that the testing concept outlined in the [first section of this blog post](https://husobee.github.io/golang/testing/unit-test/2015/06/08/golang-unit-testing.html) (this is how I got the idea for what I do here) has some utility. I promise that I'll try not to over-use it, but I strongly believe that there are cases in which this can be very useful. 

> They also make tests stateful where, if one test doesn't properly clean up after itself, the next test fails.

They show how to prevent this in the post; I forgot to add this extra part in my tests. Basically you'd do:

```
oldFactoryFunc := newRoute53Client
defer func () { newRoute53Client =  oldFactoryFunc }
newRoute53Client = func(config *config.CommandConfig) route53Client {
		return &mock
}
```

While this technique may be slightly confusing if you have never seen if before, there are cases where I think it make your code simpler and cleaner- I'll be using this in some of the other Service Discovery code. Admittedly, this PR is not the best use case for it; so I'll switch to a more conventional approach. I'll use it sparingly- when I have a strong justification to do so.

Reply:
> They show how to prevent this in the post; I forgot to add this extra part in my tests. Basically you'd do:

It still is dependent on best intentions of humans, which is the problematic bit. It's very easy to miss out on things like this as time goes on and that is primarily why I advocate for never using global variables.

> While this technique may be slightly confusing if you have never seen if before

There's also something to be said about keeping the code as simple as possible. You have to keep in mind the future committers/collaborators to a code base and the ramp up time time involved in getting some one up to speed with a code base. Generating mocks is a pretty standard thing in go and go favors code generation wherever possible. Just look at the number of workarounds this pattern needs to get around not generating mocks. Where possible, sticking to the idiomatic way of doing things has lots of value.